File: src/main/java/com/sleepkqq/sololeveling/avro/config/DefaultKafkaConfig.java
package com.sleepkqq.sololeveling.avro.config;

import io.confluent.kafka.serializers.KafkaAvroDeserializer;
import io.confluent.kafka.serializers.KafkaAvroSerializer;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.kafka.common.serialization.StringSerializer;
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory;
import org.springframework.kafka.core.ConsumerFactory;
import org.springframework.kafka.core.DefaultKafkaConsumerFactory;
import org.springframework.kafka.core.DefaultKafkaProducerFactory;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.core.ProducerFactory;

@RequiredArgsConstructor
public class DefaultKafkaConfig {

  private static final String SCHEMA_REGISTRY_URL_CONFIG = "schema.registry.url";
  private static final String SPECIFIC_AVRO_READER_CONFIG = "specific.avro.reader";

  private final String bootstrapServers;
  private final String schemaRegistryUrl;

  public <V> ProducerFactory<String, V> createProducerFactory() {
    return new DefaultKafkaProducerFactory<>(Map.of(
        ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers,
        ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class,
        ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, KafkaAvroSerializer.class,
        SCHEMA_REGISTRY_URL_CONFIG, schemaRegistryUrl
    ));
  }

  public <V> KafkaTemplate<String, V> createKafkaTemplate(
      ProducerFactory<String, V> producerFactory) {
    return new KafkaTemplate<>(producerFactory);
  }

  public <V> ConsumerFactory<String, V> createConsumerFactory(String groupId) {
    return new DefaultKafkaConsumerFactory<>(Map.of(
        ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers,
        ConsumerConfig.GROUP_ID_CONFIG, groupId,
        ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class,
        ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, KafkaAvroDeserializer.class,
        SCHEMA_REGISTRY_URL_CONFIG, schemaRegistryUrl,
        SPECIFIC_AVRO_READER_CONFIG, Boolean.TRUE.toString()
    ));
  }

  public <V> ConcurrentKafkaListenerContainerFactory<String, V> createKafkaListenerContainerFactory(
      ConsumerFactory<String, V> consumerFactory
  ) {
    var factory = new ConcurrentKafkaListenerContainerFactory<String, V>();
    factory.setConsumerFactory(consumerFactory);
    return factory;
  }
}

File: src/main/kotlin/com/sleepkqq/sololeveling/avro/config/DefaultKafkaConfig.kt
package com.sleepkqq.sololeveling.avro.config

import io.confluent.kafka.serializers.KafkaAvroDeserializer
import io.confluent.kafka.serializers.KafkaAvroSerializer
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.common.serialization.StringDeserializer
import org.apache.kafka.common.serialization.StringSerializer
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory
import org.springframework.kafka.core.*
import java.lang.Boolean
import kotlin.String

@Suppress("unused")
abstract class DefaultKafkaConfig(
	private val bootstrapServers: String,
	private val schemaRegistryUrl: String
) {

	companion object {
		private const val SCHEMA_REGISTRY_URL_CONFIG = "schema.registry.url"
		private const val SPECIFIC_AVRO_READER_CONFIG = "specific.avro.reader"
	}

	fun <V> createProducerFactory(): ProducerFactory<String, V> {
		return DefaultKafkaProducerFactory<String, V>(
			mapOf(
				ProducerConfig.BOOTSTRAP_SERVERS_CONFIG to bootstrapServers,
				ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG to StringSerializer::class.java,
				ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG to KafkaAvroSerializer::class.java,
				SCHEMA_REGISTRY_URL_CONFIG to schemaRegistryUrl
			)
		)
	}

	fun <V> createKafkaTemplate(
		producerFactory: ProducerFactory<String, V>
	): KafkaTemplate<String, V> = KafkaTemplate<String, V>(producerFactory)

	fun <V> createConsumerFactory(groupId: String): ConsumerFactory<String, V> {
		return DefaultKafkaConsumerFactory<String, V>(
			mapOf(
				ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG to bootstrapServers,
				ConsumerConfig.GROUP_ID_CONFIG to groupId,
				ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG to StringDeserializer::class.java,
				ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG to KafkaAvroDeserializer::class.java,
				SCHEMA_REGISTRY_URL_CONFIG to schemaRegistryUrl,
				SPECIFIC_AVRO_READER_CONFIG to Boolean.TRUE.toString()
			)
		)
	}

	fun <V> createKafkaListenerContainerFactory(
		consumerFactory: ConsumerFactory<String, V>
	): ConcurrentKafkaListenerContainerFactory<String, V> {
		val factory = ConcurrentKafkaListenerContainerFactory<String, V>()
		factory.consumerFactory = consumerFactory
		return factory
	}
}
